rust/plugin-lib/src/global/mod.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod view;
+mod dispatch;
+
+use std::io;
+use std::path::Path;
+
+use xi_rpc::{RpcLoop, ReadError};
+use xi_rope::rope::RopeDelta;
+use xi_core::ConfigTable;
+use xi_core::plugin_rpc::PluginEdit;
+
+use plugin_base::{Error, DataSource};
+use self::dispatch::Dispatcher;
+
+pub use self::view::View;
+
+///
+/// A generic interface for types that cache a remote document.
+///
+/// In general, users of this library should not need to implement this trait;
+/// we provide two concrete Cache implementations, [`ChunkCache`] and
+/// [`StateCache`]. If however a plugin's particular needs are not met by
+/// those implementations, a user may choose to implement their own.
+///
+/// [`ChunkCache`]: ../base_cache/struct.ChunkCache.html
+/// [`StateCache`]: ../state_cache/struct.StateCache.html
+pub trait Cache {
+    /// Create a new instance of this type; instances are created automatically
+    /// as relevant views are added.
+    fn new(buf_size: usize, rev: u64, num_lines: usize) -> Self;
+    /// Returns the line at `line_num` (zero-indexed). Returns an `Err(_)` if
+    /// there is a problem connecting to the peer, or if the requested line
+    /// is out of bounds.
+    ///
+    /// The `source` argument is some type that implements [`DataSource`]; in
+    /// the general case this is backed by the remote peer.
+    ///
+    /// [`DataSource`]: trait.DataSource.html
+    fn get_line<DS>(&mut self, source: &DS, line_num: usize) -> Result<&str, Error>
+        where DS: DataSource;
+    /// Updates the cache by applying this delta.
+    fn update(&mut self, delta: Option<&RopeDelta>, buf_size: usize,
+              num_lines: usize, rev: u64);
+    /// Flushes any state held by this cache.
+    fn clear(&mut self);
+}
+
+/// An interface for plugins.
+///
+/// Users of this library must implement this trait for some type.
+pub trait Plugin {
+    type Cache: Cache;
+
+    //TODO: async edits only; this is here for feature paritiy during initial hacking
+    /// Called when an edit has occured in the remote view. If the plugin wishes
```

Comment:
I think we wanted to move away from this, in favor of plugin edits being async, so not sure why I'm seeing it in new code. Perhaps it's related to the fact we'll need more sophisticated CRDT so a sequence of async edits appear to the plugin as if it's editing its own copy of the doc.

I guess this is more of a repackaging of existing stuff, so I'm willing to be convinced those changes should happen separately.